microcosm_flask/fields/uri_field.py

```diff
@@ -0,0 +1,93 @@
+"""
+A URI valued (and normalized) field.
+
+"""
+from operator import itemgetter
+
+from marshmallow.fields import Field, ValidationError
+from rfc3986 import uri_reference
+
+
+DEFAULT_PORTS = {
+    "http": 80,
+    "https": 443,
+}
+
+
+def normalize_uri(uri):
+    """
+    Normalize a URI (and return a string)
+
+    """
+    return normalize_uri_result(uri).unsplit()
+
+
+def normalize_uri_result(uri):
+    """
+    Normalize a URI (And return a URIResult)
+
+    """
+    ref = uri_reference(uri).normalize()
+
+    return ref._replace(
```

Comment:
I'm still not super happy with the `rfc3986` library; it doesn't quite normalize as I'd hope so I wrote some extra processing below.